sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -113,7 +114,7 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
     // literal and cast side, then process the result and swap the literal and cast again to
     // restore the original order.
     case BinaryComparison(Literal(_, literalType), Cast(fromExp, toType, _, _))
-        if canImplicitlyCast(fromExp, toType, literalType) =>
+        if needToCanonicalizeExpression(fromExp, toType, literalType) =>
```

Comment:
OK now I got it. I feel `needToCanonicalizeExpression` is confusing and updating `canImplicitlyCast` directly is fine. We may need to rename `simplifyNumericComparison` as it handles datetime as well.